TESTS/block_device/qspif/main.cpp

```diff
@@ -0,0 +1,339 @@
+    TEST_ASSERT_EQUAL(0, err);
+
+    printf("read  %0*llx:%llu ", addrwidth, block, block_size);
+    for (int i_ind = 0; i_ind < 15; i_ind++) {
+        printf("%02x", read_block[block_size - 16 + i_ind]);
+    }
+    printf("...\n");
+
+    // Find error mask for debugging
+    memset(error_mask, 0, TEST_ERROR_MASK);
+    bd_size_t error_scale = block_size / (TEST_ERROR_MASK * 8);
+
+    srand(seed);
+    for (bd_size_t i_ind = 0; i_ind < TEST_ERROR_MASK * 8; i_ind++) {
+        for (bd_size_t j_ind = 0; j_ind < error_scale; j_ind++) {
+            if ((0xff & rand()) != read_block[i_ind * error_scale + j_ind]) {
+                error_mask[i_ind / 8] |= 1 << (i_ind % 8);
+            }
+        }
+    }
+
+    printf("error %0*llx:%llu ", addrwidth, block, block_size);
+    for (int i_ind = 0; i_ind < 16; i_ind++) {
+        printf("%02x", error_mask[i_ind]);
+    }
+    printf("\n");
+
+    // Check that the data was unmodified
+    srand(seed);
+    for (bd_size_t i_ind = 0; i_ind < block_size; i_ind++) {
+        TEST_ASSERT_EQUAL(0xff & rand(), read_block[i_ind]);
+    }
+
+    err = blockD.deinit();
+    TEST_ASSERT_EQUAL(0, err);
+}
+
+
+
+static void test_qspif_thread_job(void *vBlockD/*, int thread_num*/)
+{
+    static int thread_num = 0;
+    thread_num++;
+    QSPIFBlockDevice *blockD = (QSPIFBlockDevice *)vBlockD;
+    printf("\n Thread %d Started \n", thread_num);
+
+    bd_size_t block_size = blockD->get_erase_size();
+    uint8_t *write_block = new uint8_t[block_size];
+    uint8_t *read_block = new uint8_t[block_size];
+    uint8_t *error_mask = new uint8_t[TEST_ERROR_MASK];
+    unsigned addrwidth = ceil(log(float(blockD->size() - 1)) / log(float(16))) + 1;
+
+    for (int b = 0; b < TEST_BLOCK_COUNT; b++) {
+        basic_erase_program_read_test((*blockD), block_size, write_block, read_block, error_mask, addrwidth);
+    }
+
+
+}
+
+void test_qspif_multi_threads()
+{
+
+    printf("\nTest Multi Threaded Erase/Program/Read Starts..\n");
+
+    QSPIFBlockDevice blockD(MBED_CONF_QSPIF_QSPI_IO0, MBED_CONF_QSPIF_QSPI_IO1, MBED_CONF_QSPIF_QSPI_IO2,
+                            MBED_CONF_QSPIF_QSPI_IO3, MBED_CONF_QSPIF_QSPI_CLK, MBED_CONF_QSPIF_QSPI_CS, QSPIF_POLARITY_MODE_0,
+                            MBED_CONF_QSPIF_QSPI_FREQ);
+
+    int err = blockD.init();
+    TEST_ASSERT_EQUAL(0, err);
+
+    for (unsigned atr = 0; atr < sizeof(ATTRS) / sizeof(ATTRS[0]); atr++) {
+        static const char *prefixes[] = {"", "k", "M", "G"};
+        for (int i_ind = 3; i_ind >= 0; i_ind--) {
+            bd_size_t size = (blockD.*ATTRS[atr].method)();
+            if (size >= (1ULL << 10 * i_ind)) {
+                printf("%s: %llu%sbytes (%llubytes)\n",
+                       ATTRS[atr].name, size >> 10 * i_ind, prefixes[i_ind], size);
+                break;
+            }
+        }
+    }
+
+    rtos::Thread first_qspif_bd_thread;
+    rtos::Thread second_qspif_bd_thread;
+    rtos::Thread third_qspif_bd_thread;
+
```

Comment:
Why not use an array and loop here?

Reply:
I thought 3 would be enough (-:
I'll upgrade to an array